preserve spans in macros to generate better rustdoc [src] links

mystor opened the pull request:
Make rustdoc generate the correct `[src]` links by perserving spans in our macros.

`__check` makes sure you don't write the wrong keywords.

dtolnay commented:
Is this for #304? I think `$(#[$attr:meta])* $pub:ident` is an ambiguity prior to Rust 1.20.

mystor commented:
Yeah, this is for #304. Shucks re: the ambiguity thing :cry: 

I'll have to fiddle around a bit more to see if I can work around that while still preserving the spans...

mystor commented:
@dtolnay With 1.0 bumping the min rustc version this may be worth resurrecting :-)

mystor commented:
I updated it, added some fallback logic. I had to do some additional minor changes to ensure that as many items as possible get correctly spanned, because if rustc sees that either the first or last span in the item is attributed to a macro, it attributes the entire span to the macro.

Unfortunately, I was unable to get the spans generated the outer enum from `ast_enum_of_structs!` to be correct (though I was able to get correct spans for the inner generated structs). I added a comment to the code to explain why (copied below): 

```
// Unfortunately, at this time, we can't make the generated enum here have the
// correct span. The way that the span for the overall enum decl is calculated
// is using `lo.to(prev_span)` [1].
//
// If the beginning and ending spans of the enum item don't have the same macro
// context, fallback code is run [2]. This code chooses the span within the
// macro context to avoid firing spurious diagnostics [3].
//
// This means `[src]` links in rustdoc will point to the macro, instead of the
// actual declaration, if either the first or last token of our declaration has
// a span from the macro. With `macro_rules!` there is no way to preserve the
// span of a `{}` block while changing its contents, so we're forced to have the
// span of our final token point to our macro.
//
// [1]: https://github.com/rust-lang/rust/blob/9a90d03ad171856dc016c2dcc19292ec49a8a26f/src/libsyntax/parse/parser.rs#L7377
// [2]: https://github.com/rust-lang/rust/blob/9a90d03ad171856dc016c2dcc19292ec49a8a26f/src/libsyntax_pos/lib.rs#L467-L478
// [3]: https://github.com/rust-lang/rust/pull/47942
```
